File: 7-2.py
#!/usr/bin/env python

from collections import deque
import re

class Bag:
  def __init__(self, colour):
    self.colour = colour
    self.parents = []
    self.children = {}

  def add_parent(self, colour):
    self.parents.append(colour)

  def add_child(self, colour, num):
    self.children[colour] = num

  def visit(self):
    num = 1
    for colour, count in self.children.items():
      num += bags[colour].visit() * count
    return num

bags = {}

def get_bag(colour):
    b = bags.get(colour, None)
    if not b:
        b = Bag(colour)
        bags[colour] = b
    return b

for edge in open('7.input').readlines():
    mb = re.search('(?P<source>[ a-z]*) bags contain (?P<sinks>[, a-z0-9]*)', edge)
    b = get_bag(mb['source'])
    bags[mb['source']] = b
    if mb['sinks'] == 'no other bags':
        b.children = {}
    else:
        for c in mb['sinks'].split(','):
            mc = re.search(' ?(?P<num>[0-9]*) (?P<colour>[ a-z]*) bag(s?)', c)
            b.add_child(mc['colour'], int(mc['num']))
            get_bag(mc['colour']).add_parent(mb['source'])

print(f"{bags['shiny gold'].visit() - 1}")

File: 7-2.rb
#!/usr/bin/env ruby

class Bag
  attr_accessor :colour

  def initialize(colour)
    @colour = colour
    @parents = []
    @children = {}
  end

  attr_accessor :parents

  def add_parent(colour)
    @parents.push colour
  end

  attr_accessor :children

  def add_child(colour, num)
    @children[colour] = num
  end

  def visit
    num = 1
    @children.each do |colour, count|
      num += $bags[colour].visit * count
    end
    num
  end
end

$bags = {}

def get_bag(colour)
  $bags.fetch(colour) { |c| $bags[c] = Bag.new(c) }
end

File.read('7.input').lines.each do |edge|
  /(?<source>[ a-z]*) bags contain (?<sinks>[, a-z0-9]*)/.match(edge) do |mb|
    b = get_bag(mb['source'])
    $bags[mb['source']] = b
    if mb['sinks'] == 'no other bags'
      b.children = []
    else
      mb['sinks'].split(',') do |c|
        mc = / ?(?<num>[0-9]*) (?<colour>[ a-z]*) bag(s?)/.match(c)
        b.add_child(mc['colour'], mc['num'].to_i)
        get_bag(mc['colour']).add_parent mb['source']
      end
    end
  end
end

print "#{$bags['shiny gold'].visit - 1}\n"
